src/SFML/Audio/SoundFileReaderMp3.cpp

```diff
@@ -0,0 +1,134 @@
+////////////////////////////////////////////////////////////
+
+////////////////////////////////////////////////////////////
+// Headers
+////////////////////////////////////////////////////////////
+#define MINIMP3_IMPLEMENTATION // Minimp3 control define, places implementation in this file.
+#include <SFML/Audio/SoundFileReaderMp3.hpp>
+#include <SFML/System/MemoryInputStream.hpp>
+#include <SFML/System/Err.hpp>
+#include <algorithm>
+
+
+namespace
+{
+std::size_t readCallback(void* ptr, std::size_t size, void* data)
+{
+    sf::InputStream* stream = static_cast<sf::InputStream*>(data);
+    return static_cast<std::size_t>(stream->read(ptr, size));
+}
+
+int seekCallback(uint64_t offset, void* data)
+{
+    sf::InputStream* stream = static_cast<sf::InputStream*>(data);
+    sf::Int64 position = stream->seek(offset);
+    return position < 0 ? -1 : 0;
+}
+}
+
+namespace sf
+{
+namespace priv
+{
+////////////////////////////////////////////////////////////
+bool SoundFileReaderMp3::check(InputStream& stream)
+{
+    Uint8 header[10];
+    if (stream.read(header, sizeof(header)) < sizeof(header))
+        return false;
+    if (!memcmp(header, "ID3", 3) && !((header[5] & 15) || (header[6] & 0x80) || (header[7] & 0x80) || (header[8] & 0x80) || (header[9] & 0x80)))
+        return true;
+    if (hdr_valid(header))
+        return true;
+    return false;
+}
+
+
+////////////////////////////////////////////////////////////
+SoundFileReaderMp3::SoundFileReaderMp3() :
+m_numSamples(0),
+m_position(0)
+{
+    m_io.read = readCallback;
+    m_io.seek = seekCallback;
+}
+
+
+////////////////////////////////////////////////////////////
+SoundFileReaderMp3::~SoundFileReaderMp3()
+{
+    close();
+}
+
+
+////////////////////////////////////////////////////////////
+bool SoundFileReaderMp3::open(InputStream& stream, Info& info)
+{
+    // Init IO callbacks
+    m_io.read_data = &stream;
+    m_io.seek_data = &stream;
+
+    // Init mp3 decoder
+    mp3dec_ex_open_cb(&m_decoder, &m_io, MP3D_SEEK_TO_SAMPLE);
+    if (!m_decoder.samples)
+        return false;
+
+    // Retrieve the music attributes
+    info.channelCount = m_decoder.info.channels;
+    info.sampleRate   = m_decoder.info.hz;
+    info.sampleCount  = m_decoder.samples;
+
+    m_numSamples      = info.sampleCount;
+    return true;
+}
+
+
+////////////////////////////////////////////////////////////
+void SoundFileReaderMp3::seek(Uint64 sampleOffset)
+{
+    m_position = std::min(sampleOffset, m_numSamples);
+    mp3dec_ex_seek(&m_decoder, m_position);
+}
+
+
+////////////////////////////////////////////////////////////
+Uint64 SoundFileReaderMp3::read(Int16* samples, Uint64 maxCount)
+{
+    Uint64 toRead = std::min(maxCount, m_numSamples - m_position);
+    toRead = mp3dec_ex_read(&m_decoder, samples, toRead);
+    m_position += toRead;
+    return toRead;
+}
+
+
+////////////////////////////////////////////////////////////
+void SoundFileReaderMp3::close()
+{
+    mp3dec_ex_close(&m_decoder);
```

Comment:
It seems the only time `close()` is invoked is in the destructor, and the method is private.

Would it make sense to inline it directly (i.e. calling `mp3dec_ex_close(&m_decoder);` in the destructor)?
This would avoid accidentally calling `close()` twice if the implementation were to change in the future.

Reply:
Done. Updated the PR.

Reply:
This method can then be removed.

Reply:
Done. Updated the PR.